### DISCUSSION TITLE
How do I leverage existing KafkaTemplate from Producer config in RetryTopicConfiguration?

### DISCUSSION THREAD

**Discussion by hariraogotit** [spring-projects/spring-kafka#2637] [Q&A] [answered]
Team,

  With reference to https://docs.spring.io/spring-kafka/reference/html/#using-retrytopicconfiguration-beans I would like to know how I can leverage existing KafkaTemplate (getSdKafkaTemplate) configured in my KafkaProducer config?

I have the following in MyRetryConfig.java that is annotated with @Configuration

```
@Bean
 public RetryTopicConfiguration myRetryTopic(KafkaTemplate<String, byte[]> kafkaTemplate) {
   return RetryTopicConfigurationBuilder
           .newInstance()
           .exponentialBackoff(3000, 2, 100000)
           .maxAttempts(5)
           .retryTopicSuffix("-asl-nm.retry")
           .dltSuffix("-asl-nm.dlt")
           .create(kafkaTemplate);
 }
```
In KafkaPublisherConfiguration.java that is also annotated with @Configuration has the following 

```
@Bean
  KafkaTemplate<String, byte[]>  getSdKafkaTemplate(
      @Value("${spring.kafka.bootstrap-servers}") String sdServers) {
    return getProducerTemplate(sdServers, bufferMemory, securityProtocol);
  }
```
I want to make sure that the retry uses the configured producer. So my understanding is in lieu of `KafkaTemplate<String, byte[]> kafkaTemplate` I need to pass `KafkaTemplate<String, byte[]> getSdKafkaTemplate` in `RetryTopicConfiguration myRetryTopic(` to leverage the existing configured KafkaProducer? Kindly let me know if my understanding is correct?

**Comment by gpando** [comment, marked as answer]
Use a qualifier:


```
 @Bean(name = "mykafkatemplate")
 KafkaTemplate<String, byte[]>  getSdKafkaTemplate(
      @Value("${spring.kafka.bootstrap-servers}") String sdServers) {   
      return getProducerTemplate(sdServers, bufferMemory, securityProtocol);   
  }   

```

```
 @bean
 public RetryTopicConfiguration myRetryTopic(@Qualifier("mykafkatemplate") KafkaTemplate<String, byte[]> kafkaTemplate) {
  ...

```

**Reply by garyrussell** [reply]
Or use the bean name as the parameter name:

```java
public RetryTopicConfiguration myRetryTopic(KafkaTemplate<String, byte[]> mykafkatemplate) {
  ...
```

**Comment by hariraogotit** [comment]
Lovely thanks @garyrussell and @gpando . Will try the suggestions and marked as answered.

**Reply by hariraogotit** [reply]
Thank you @garyrussell  @gpando  for your time.
